lte/gateway/c/oai/tasks/ngap/ngap_state_manager.cpp

```diff
@@ -0,0 +1,183 @@
+  AssertFatal(
+      is_initialized,
+      "NgapStateManager init() function should be called to initialize state.");
+
+  if (state_cache_p == nullptr) {
+    return;
+  }
+
+  int i;
+  hashtable_rc_t ht_rc;
+  hashtable_key_array_t* keys;
+  sctp_assoc_id_t assoc_id;
+  gnb_description_t* gnb;
+
+  keys = hashtable_ts_get_keys(&state_cache_p->gnbs);
+  if (!keys) {
+    OAILOG_DEBUG(LOG_NGAP, "No keys in the amf hashtable");
+  } else {
+    for (i = 0; i < keys->num_keys; i++) {
+      assoc_id = (sctp_assoc_id_t) keys->keys[i];
+      ht_rc    = hashtable_ts_get(
+          &state_cache_p->gnbs, (hash_key_t) assoc_id, (void**) &gnb);
+      AssertFatal(ht_rc == HASH_TABLE_OK, "eNB UE id not in assoc_id");
+      AssertFatal(ht_rc == HASH_TABLE_OK, "eNB UE id not in assoc_id");
+    }
+    FREE_HASHTABLE_KEY_ARRAY(keys);
+  }
+
+  if (hashtable_ts_destroy(&state_cache_p->gnbs) != HASH_TABLE_OK) {
+    OAI_FPRINTF_ERR("An error occurred while destroying s1 eNB hash table");
+  }
+  if (hashtable_ts_destroy(&state_cache_p->amfid2associd) != HASH_TABLE_OK) {
+    OAI_FPRINTF_ERR("An error occurred while destroying assoc_id hash table");
+  }
+  if (hashtable_ts_destroy(state_ue_ht) != HASH_TABLE_OK) {
+    OAI_FPRINTF_ERR("An error occurred while destroying assoc_id hash table");
+  }
+  free_wrapper((void**) &state_cache_p);
+
+  clear_ngap_imsi_map();
+}
+
+int NgapStateManager::read_ue_state_from_db() {
+  if (!persist_state_enabled) {
+    return RETURNok;
+  }
+  auto keys = redis_client->get_keys("IMSI*" + task_name + "*");
+
+  for (const auto& key : keys) {
+    Ngap_UeDescription ue_proto = Ngap_UeDescription();
+    m5g_ue_description_t* ue_context =
+        (m5g_ue_description_t*) calloc(1, sizeof(m5g_ue_description_t));
+    if (redis_client->read_proto(key.c_str(), ue_proto) != RETURNok) {
+      return RETURNerror;
+    }
+
+    NgapStateConverter::proto_to_ue(ue_proto, ue_context);
+
+    hashtable_ts_insert(
+        state_ue_ht, ue_context->comp_ngap_id, (void*) ue_context);
+    OAILOG_DEBUG(log_task, "Reading UE state from db for %s", key.c_str());
+  }
+  return RETURNok;
+}
+
+void NgapStateManager::create_ngap_imsi_map() {
+  ngap_imsi_map_ = (ngap_imsi_map_t*) calloc(1, sizeof(ngap_imsi_map_t));
+
+  ngap_imsi_map_->amf_ue_id_imsi_htbl =
+      hashtable_uint64_ts_create(max_ues_, nullptr, nullptr);
+
+  oai::NgapImsiMap imsi_proto = oai::NgapImsiMap();
+  redis_client->read_proto(NGAP_IMSI_MAP_TABLE_NAME, imsi_proto);
+
+  NgapStateConverter::proto_to_ngap_imsi_map(imsi_proto, ngap_imsi_map_);
+}
+
+void NgapStateManager::clear_ngap_imsi_map() {
+  if (!ngap_imsi_map_) {
+    return;
+  }
+  hashtable_uint64_ts_destroy(ngap_imsi_map_->amf_ue_id_imsi_htbl);
+
+  free_wrapper((void**) &ngap_imsi_map_);
+}
+
+ngap_imsi_map_t* NgapStateManager::get_ngap_imsi_map() {
+  return ngap_imsi_map_;
+}
+
+void NgapStateManager::put_ngap_imsi_map() {
+  oai::NgapImsiMap imsi_proto = oai::NgapImsiMap();
```

Comment:
Wrtite to Redis only if persist_state_enabled is true
```suggestion
  if (!persist_state_enabled) {
  return;
 }
  oai::NgapImsiMap imsi_proto = oai::NgapImsiMap();
```